tests/test_serialdriver.py

```diff
@@ -23,14 +23,20 @@ def test_write(self, target, serial_port, mocker):
         serial_mock.return_value.open.assert_called_once_with()
         serial_mock.return_value.write.assert_called_once_with(b"testdata")
 
-    def test_read(self, target, serial_port, mocker):
+    @pytest.mark.parametrize("param", [[1, 1, False, 1],  # old test case
+                                       [3, 2, False, 3], [3, 2, True, 2], [1, 2, True, 2]])
+    # param = [size, in_waiting, read_block, out]
+    def test_read(self, target, serial_port, mocker, param):
         serial_mock = mocker.patch('serial.Serial')
-        serial_mock.return_value.in_waiting = 0
+        serial_mock.return_value.in_waiting = param[1]
         s = SerialDriver(target, "serial")
         target.activate(s)
-        s.read()
+        s.read(size=param[0], read_block=param[2])
         serial_mock.return_value.open.assert_called_once_with()
-        serial_mock.return_value.read.assert_called_once_with(1)
+        if not param[2]:
```

Comment:
I would split this test into two tests, so this if/else is not needed.

Reply:
I refined the testing such that no if/else condition is needed.
The test now asserts the number of read function calls and the returned values.